### ISSUE TITLE
Assigning a custom role to a new user fails (400) UNSUPPORTED_ROLE

### ISSUE THREAD

**Issue by mgingios** [mongodb/terraform-provider-mongodbatlas#1522]
### Terraform CLI and Terraform MongoDB Atlas Provider Version
```hcl
terraform version     
Terraform v1.6.1
on darwin_amd64
+ provider registry.terraform.io/hashicorp/random v3.5.1
+ provider registry.terraform.io/hashicorp/time v0.9.1
+ provider registry.terraform.io/mongodb/mongodbatlas v1.12.1
```

### Terraform Configuration File 
```hcl
resource "random_string" "db-user-password" {
  length           = 64
  upper            = true
  numeric          = true
  lower            = true
  special          = true
  override_special = "_%@"
}

locals {
  client_code     = "abcd"
  collection_name = "projects"
  database_name   = "${local.client_code}-db"
  role_name       = "${local.client_code}-role"
  username        = "${local.client_code}"
}

resource "mongodbatlas_project" "csc" {
  name   = var.mongodb_atlas_project_name
  org_id = var.mongodb_atlas_org_id
}

resource "mongodbatlas_custom_db_role" "test_role" {
  project_id = mongodbatlas_project.csc.id
  role_name  = "testdatabaserole"
  
  actions {
    action = "CREATE_COLLECTION"
    resources {
      collection_name = ""
      database_name   = "testdatabase"
    }
  }
  actions {
    action = "CREATE_INDEX"
    resources {
      collection_name = ""
      database_name   = "testdatabase"
    }
  }
  actions {
    action = "INSERT"
    resources {
      collection_name = ""
      database_name   = "testdatabase"
    }
  }
  actions {
    action = "FIND"
    resources {
      collection_name = ""
      database_name   = "testdatabase"
    }
  }
  actions {
    action = "REMOVE"
    resources {
      collection_name = ""
      database_name   = "testdatabase"
    }
  }
  actions {
    action = "UPDATE"
    resources {
      collection_name = ""
      database_name   = "testdatabase"
    }
  }

  inherited_roles {
    role_name     = "readWrite"
    database_name = "testdatabase"
  }
}

# The following delay was suggested by the MongoDB Community Supporters.
# The idea is to inject a little time between the creation of the custom role and
# the creation of the new user that will be assigned the custom role.
resource "time_sleep" "wait_30_seconds" {
  depends_on = [mongodbatlas_custom_db_role.test_role]

  create_duration = "30s"
}

resource "mongodbatlas_database_user" "test_user1" {
  username           = "testuser1"
  password           = "testpassword1"
  project_id         = mongodbatlas_project.csc.id
  auth_database_name = "admin"

  roles {
    role_name     = "testdatabaserole"
    database_name = "testdatabase"
  }

  depends_on = [
    mongodbatlas_custom_db_role.test_role
  ]
}


output "next_steps" {
  value = "The database access role and user has been created at MongoDB Atlas.  The password can be obtained by executing: 'terraform output -raw client-password'.  You need to create a database named '${local.database_name}' and a collection within the database named 'projects'."
}
```

### Steps to Reproduce
1. `terraform init`
2. `terraform apply`

### Expected Behavior
The custom role is created, and a new user is created and assigned to the custom role.

### Actual Behavior

- If I create the custom role and not create the user, it works properly.
- If I create the user without a custom role, but use an inherited one, it works properly.
- If I create the custom role with a dependent delay, as proffered in the closed ticket, and create the user, it fails with "400 (request "UNSUPPORTED_ROLE") The provided role is not supported."  The custom role exists, but the user was not created.
- If I use the MongoDB Atlas UI, I can create a new user and assign the custom role to the new user.
- If I comment out the creation of the new user and assignment to the custom role to only create the new custom role and 'terraform apply'.  The new role is created.  If I uncomment the creation and assignment of a new user to the role and 'terraform apply', I get the same UNSUPPORTED_ROLE error.

### Debug Output 
mongodbatlas_custom_db_role.test_role: Creating...
mongodbatlas_custom_db_role.test_role: Creation complete after 4s [id=cHJvamVjdF9pZA==:NWNiNjZhOTYwMTRiNzY5ODRlMGY4NWE1-cm9sZV9uYW1l:dGVzdGRhdGFiYXNlcm9sZQ==]
mongodbatlas_database_user.test_user1: Creating...
╷
│ Error: error during database user creation
│ 
│   with mongodbatlas_database_user.test_user1,
│   on main.tf line 81, in resource "mongodbatlas_database_user" "test_user1":
│   81: resource "mongodbatlas_database_user" "test_user1" {
│ 
│ POST https://cloud.mongodb.com/api/atlas/v1.0/groups/5cb66a96014b76984e0f85a5/databaseUsers: 400 (request "UNSUPPORTED_ROLE") The provided role is not supported.

### Crash Output
N/A

### Additional Context
NONE

### References
I am experiencing exactly the same issue reported in the following link, which was closed in 2020.
https://github.com/mongodb/terraform-provider-mongodbatlas/issues/273#issue-664516965

**Comment by AgustinBettati** [comment]
Hello @mgingios.

Thank you for the details provided. I was able to reproduce the issue, having the custom db role successfully created and not being able to use within the mongodbatlas_database_user resource. I will get back to you when we have new findings or a fix from our end.

**Comment by mrbcmorris** [comment]
:wave: Hey @mgingios ,

I was running into this exact problem yesterday.  What I found out is custom roles, when being attached to users, always get attached as the `admin` database.  Once I made this change, I was able to get my terraform to apply cleanly.

We should likely submit a PR to the TF docs because it seems like a lot of folks are struggling with this, but regardless just wanted to share what I learned.

**Comment by mgingios** [comment]
Hi @AgustinBettati & @mrbcmorris, Thanks for validating this issue!  

So, I changed the following assignment to use the 'admin' database rather than the one that the user will access for data.

mongodbatlas_database_user.roles.database_name = "abcd"

to

mongodbatlas_database_user.roles.database_name = "admin"

After purging the user and role from MongoDB Atlas, I applied the TF script and it worked!  So, I can confirm @mrbcmorris' solution.  The comment that I placed in the script may be useful in the documentation or examples.

```
resource "mongodbatlas_database_user" "csc" {
  username           = local.username
  password           = random_string.db-user-password.result
  project_id         = mongodbatlas_project.csc.id
  depends_on = [mongodbatlas_custom_db_role.csc]
  auth_database_name = "admin"
  # The database and collection do not need to exist in the cluster before creating the user.
  roles {
    role_name     = local.role_name
    # The roles.database_name must be the 'admin' database when assigning a custom role to the user.  
    # The user accesses the database defined in the custom role.
    database_name = "admin"
  }
  scopes {
    name = "VicDashboard"
    type = "CLUSTER"
  }
}
```

Thanks fo your help with this problem, guys!

Mike

**Comment by AgustinBettati** [comment]
@mrbcmorris Thank you for the quick feedback and @mgingios for verifying the solution.
As a follow up we will look into adjusting the documentation to avoid confusions for this use case.

**Comment by ezloj** [comment]
Hello,

I am also having this issue. It is strange that the input for the user with a custom role is `admin` database. If I understand correctly, a role's database/collection options will dictate the database access permissions instead of the user's database option. So in this case the fact that a user resource is provided with `admin` for `database_name` argument is misleeading as it will have no effect.

Is my understanding correct? If so, then it's not the docs that have to be updated. Instead, the provider should be updated to have `database_name` argument for the user resource be optional when supplying a custom role. Additionally, API response should provide more info about what is happening rather than plain `400: UNSUPPORTED_ROLE`

Jan

**Comment by maastha** [comment]
Hi @ezloj 
Thank you for your insights on this.

Upon reproducing the error I noticed we do include the details in the error message:
```
│ Error: error during database user creation
│ 
│   with mongodbatlas_database_user.test_user1,
│   on database_user.tf line 85, in resource "mongodbatlas_database_user" "test_user1":
│   85: resource "mongodbatlas_database_user" "test_user1" {
│ 
│ https://cloud-dev.mongodb.com/api/atlas/v2/groups/65a7b75d3da7844889b3abb2/databaseUsers POST: HTTP 400 Bad Request (Error code: "UNSUPPORTED_ROLE") Detail: The provided
│ role is not supported. Reason: Bad Request. Params: [Custom role testdatabaserole must scoped to admin database]
```

As for your comments regarding role's database/collection options to dictate the database access permissions, I have reached out to the internal team that owns this API to get further details

**Comment by maastha** [comment]
Sharing additional context regarding roles for future readers.
`databaseName` is a [required attribute](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/v2/#tag/Database-Users/operation/createDatabaseUser) for our API when assigning roles to a database user. 

While custom roles are defined with the `admin` database by Atlas on the backend, there are other built-in MongoDB roles, like `readWrite`, that do not need to be scoped to the `admin` database when assigned to a user, and can be scoped to any database name.
Given the above, for API consistency purposes users are required to set `databaseName` to `admin` for custom roles. 

Updating resource documentation to reflect the same in https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1862

**Comment by AgustinBettati** [comment]
Closing as change was merged in 1.15.0 release.
